### DISCUSSION TITLE
Could I have some help with a red-green color blind shader?

### DISCUSSION THREAD

**Discussion by lividhen** [loqusion/hyprshade#25] [Q&A] [answered]
This is what I have so far, I dont know any glsl pretty much at all. I also don't fully understand what im doing here, this has been cobbled together from a bunch of different sources. It does not look right at all. Completely blown out and some of the colors such as purple cant be seen on a white background. I have just been messing with the values for the last little bit but nothing looks right. If someone could help explain why that is happening that'd be great 😅
```
precision highp float;
varying vec2 v_texcoord;
uniform sampler2D tex;

const float gamma = 1.0;

void main() {
  vec4 pixColor = texture2D(tex, v_texcoord);

  //color matrix
  const mat4 colorMatrix = mat4(
    vec4(0.92, 0.44, 0.11, 0.0),  //r
    vec4(0.8, 0.7, 0.42, 0.0),   //g
    vec4(0.0, 0.15, 1.25, 0.0),   // b
    vec4(0.0, 0.0, 0.0, 1.0) //a
  );
  vec4 transformedColor = colorMatrix * pixColor;
  transformedColor.rgb = pow(transformedColor.rgb, vec3(gamma));
  gl_FragColor =  vec4(transformedColor.rgb, transformedColor.a);
```

**Comment by loqusion** [comment]
I also know next to nothing about glsl 😅. Does [this](https://godotshaders.com/shader/colorblindness-correction-shader/) work well for you? (You may have to change the `mode` or augment in other ways)

I'd be interested in adding some shaders (+ updating the readme) that assist those with color blindness so that no one has to reinvent the wheel.

**Reply by lividhen** [reply]
I will try that and see how it works. I have been searching for a color blind shader for hyprland for a while but haven't found anything so adding it to the readme and making it more accessible to people would be awesome!

**Comment by lividhen** [comment, marked as answer]
Got it working thanks to the resource you linked @loqusion !
Here is what I ended up with. If I could get feedback from people with different types or more intense red green color blindness that'd be great as mine is only very slight.

<details>
    <summary>Shader</summary>  

```
//Color blind shader for hyprland.
//Based off the shader :  https://godotshaders.com/shader/colorblindness-correction-shader/.

precision highp float;
varying vec2 v_texcoord;
uniform sampler2D tex;

//Intensity of filter (1.0 - 0.0)
const float intensity = 1.0;

// Color correction mode
// 0 - Protanopia
// 1 - Deutranopia
// 2 - Tritanopia
const int mode = 1;

void main()
{
  vec4 pixColor = texture2D(tex, v_texcoord);

  float L = (17.8824 * pixColor.r) + (43.5161 * pixColor.g) + (4.11935 * pixColor.b);
  float M = (3.45565 * pixColor.r) + (27.1554 * pixColor.g) + (3.86714 * pixColor.b);
  float S = (0.0299566 * pixColor.r) + (0.184309 * pixColor.g) + (1.46709 * pixColor.b);

  float l, m, s;
  if (mode == 0) //Protanopia
  {
    l = 0.0 * L + 2.02344 * M + -2.52581 * S;
    m = 0.0 * L + 1.0 * M + 0.0 * S;
    s = 0.0 * L + 0.0 * M + 1.0 * S;
  }
  
  if (mode == 1) //Deuteranopia
  {
    l = 1.0 * L + 0.0 * M + 0.0 * S;
    m = 0.494207 * L + 0.0 * M + 1.24827 * S;
    s = 0.0 * L + 0.0 * M + 1.0 * S;
  }

  if (mode == 2) //Tritanopia
  {
    l = 1.0 * L + 0.0 * M + 0.0 * S;
    m = 0.0 * L + 1.0 * M + 0.0 * S;
    s = -0.395913 * L + 0.801109 * M + 0.0 * S;
  }

  vec4 error;
  error.r = (0.0809444479 * l) + (-0.130504409 * m) + (0.116721066 * s);
  error.g = (-0.0102485335 * l) + (0.0540193266 * m) + (-0.113614708 * s);
  error.b = (-0.000365296938 * l) + (-0.00412161469 * m) + (0.693511405 * s);
  error.a = 1.0;
  vec4 diff = pixColor - error;
  vec4 correction;
  correction.r = 0.0;
  correction.g = (diff.r * 0.7) + (diff.g * 1.0);
  correction.b = (diff.r * 0.7) + (diff.b * 1.0);
  correction = pixColor + correction;
  correction.a = pixColor.a * intensity;

  gl_FragColor = correction;
}
```  

</details>

**Comment by loqusion** [comment]
@lividhen Do you think it would be useful for hyprshade to provide configurable intensity, so that people don't have to modify the shader by hand? There could be a field in `hyprland.toml`, e.g.:

```toml
[[shades]]
name = "color-filter"
[shades.config]
type = "protanopia" # "protanopia", "deuteranopia", "tritanopia"
intensity = 1.0     # 0.0 - 1.0
```

(I also added configurable `type` here to avoid the need to include multiple shader files)

On a separate note, do you have any suggestions for what to name the shader?

**Reply by lividhen** [reply]
Yeah having a config would be great! Don't really have a good name besides color blind mode or color blind shader. Probably doesn't need to be much more than that.

**Comment by Neintonine** [comment]
Hei, really nice shader.
I just noticed that, that the intensity doesn't work.
From my, honestly not much, testing, the Alpha channel is pretty much ignored, when handling the shader. So multipling anything with it doesn't change much. On Godot, it does affect the result, since the alpha is used to lerp between the original and changed value.

I modified it so it in corporates the intensity correctly. Since the 100% was a bit much for me :)
```glsl
//Color blind shader for hyprland.
//Based off the shader :  https://godotshaders.com/shader/colorblindness-correction-shader/.

precision highp float;
varying vec2 v_texcoord;
uniform sampler2D tex;

//Intensity of filter (1.0 - 0.0)
const float intensity = 1.0;

// Color correction mode
// 0 - Protanopia
// 1 - Deutranopia
// 2 - Tritanopia
const int mode = 1;

void main()
{
  vec4 pixColor = texture2D(tex, v_texcoord);

  float L = (17.8824 * pixColor.r) + (43.5161 * pixColor.g) + (4.11935 * pixColor.b);
  float M = (3.45565 * pixColor.r) + (27.1554 * pixColor.g) + (3.86714 * pixColor.b);
  float S = (0.0299566 * pixColor.r) + (0.184309 * pixColor.g) + (1.46709 * pixColor.b);

  float l, m, s;
  if (mode == 0) //Protanopia
  {
    l = 0.0 * L + 2.02344 * M + -2.52581 * S;
    m = 0.0 * L + 1.0 * M + 0.0 * S;
    s = 0.0 * L + 0.0 * M + 1.0 * S;
  }
  
  if (mode == 1) //Deuteranopia
  {
    l = 1.0 * L + 0.0 * M + 0.0 * S;
    m = 0.494207 * L + 0.0 * M + 1.24827 * S;
    s = 0.0 * L + 0.0 * M + 1.0 * S;
  }

  if (mode == 2) //Tritanopia
  {
    l = 1.0 * L + 0.0 * M + 0.0 * S;
    m = 0.0 * L + 1.0 * M + 0.0 * S;
    s = -0.395913 * L + 0.801109 * M + 0.0 * S;
  }

  vec4 error;
  error.r = (0.0809444479 * l) + (-0.130504409 * m) + (0.116721066 * s);
  error.g = (-0.0102485335 * l) + (0.0540193266 * m) + (-0.113614708 * s);
  error.b = (-0.000365296938 * l) + (-0.00412161469 * m) + (0.693511405 * s);
  error.a = 1.0;
  vec4 diff = pixColor - error;
  vec4 correction;
  correction.r = 0.0;
  correction.g = (diff.r * 0.7) + (diff.g * 1.0);
  correction.b = (diff.r * 0.7) + (diff.b * 1.0);
  correction = pixColor + (correction * intensity);
  correction.a = pixColor.a;

  gl_FragColor = correction;
}
```

**Reply by Neintonine** [reply]
Ok. Me being a dumbie, I only installed the stable version, not the git version.
For those stumbling across this... You may want to switch to git version, it has this already included as the "color-filter".
And it has the intensity (or Strength in that case) correctly implermented.

If you can't or don't want, for what ever reason, this may help you to get the same result. :)
